File: packages/kits/LocusKit/Sources/LocusKit/RecallRating.swift
import Foundation
import PersistenceKit

/// One drawer's Bradley-Terry rating from the end-of-day tournament.
///
/// Persisted in the `recall_ratings` table (`drawer_id TEXT PRIMARY KEY,
/// rating REAL, contests INTEGER, updated_at TEXT`). `rating` is the
/// estimator's log-strength for the drawer; `contests` is the running count
/// of preference observations the drawer has taken part in across every
/// tournament pass; `updatedAt` is the tournament clock at the last write,
/// stored as ISO8601 TEXT per the fleet date-storage rule.
public struct RecallRating: Sendable, Equatable {
    /// The rated drawer's identifier (`recall_ratings.drawer_id`).
    public let drawerID: String

    /// Bradley-Terry log-strength for the drawer.
    public var rating: Double

    /// Number of preference observations the drawer has appeared in.
    public var contests: Int

    /// Clock of the tournament pass that last wrote this row.
    public var updatedAt: Date

    /// Designated initializer.
    public init(drawerID: String, rating: Double, contests: Int, updatedAt: Date) {
        self.drawerID = drawerID
        self.rating = rating
        self.contests = contests
        self.updatedAt = updatedAt
    }
}

public extension RecallRating {
    /// The rating ledger's table name.
    static let tableName = "recall_ratings"

    /// The rating ledger's schema, declared through PersistenceKit's schema
    /// ladder so the DDL is identical on every backend. This is the ONE
    /// declaration of `recall_ratings` in the Swift port: `DrawerStore`
    /// applies it on first use (a fresh estate is stamped at the current
    /// format without running the migration capsules, so no capsule ever
    /// creates the table there) and the 1.8 → 1.9 estate-format capsule
    /// applies the same value on populated estates. Same kit id, version
    /// and column set on both paths, so `migrate(to:)` is a no-op once
    /// either has created the table. Mirrors Rust
    /// `recall_rating::recall_ratings_schema()`.
    static let schema = SchemaDeclaration(
        kitID: "GLKRecallRatings",
        version: 1,
        tables: [
            TableDeclaration(
                name: tableName,
                columns: [
                    .text("drawer_id", nullable: false),
                    .float("rating", nullable: false),
                    .int("contests", nullable: false),
                    .timestamp("updated_at", nullable: false),
                ],
                primaryKey: ["drawer_id"]
            )
        ]
    )
}

File: packages/kits/LocusKit/rust/src/recall_rating.rs
//! RecallRating — one drawer's Bradley-Terry rating from the end-of-day
//! tournament. Ports `RecallRating.swift`.
//!
//! Persisted in the `recall_ratings` table (`drawer_id TEXT PRIMARY KEY,
//! rating REAL, contests INTEGER, updated_at TEXT`). `rating` is the
//! estimator's strength for the drawer; `contests` is the running count of
//! preference observations the drawer has taken part in across every
//! tournament pass; `updated_at` is the tournament clock at the last write,
//! stored as ISO8601 TEXT per the fleet date-storage rule.

use persistence_kit::{ColumnDeclaration, SchemaDeclaration, TableDeclaration};

/// Name of the rating ledger table.
pub const RECALL_RATINGS_TABLE: &str = "recall_ratings";

/// One drawer's Bradley-Terry rating from the end-of-day tournament.
#[derive(Debug, Clone, PartialEq)]
pub struct RecallRating {
    /// The rated drawer's identifier (`recall_ratings.drawer_id`).
    pub drawer_id: String,

    /// Bradley-Terry strength for the drawer.
    pub rating: f64,

    /// Number of preference observations the drawer has appeared in.
    pub contests: i64,

    /// Clock of the tournament pass that last wrote this row, as an
    /// ISO8601 string (TEXT in SQLite).
    pub updated_at: String,
}

impl RecallRating {
    /// Create a rating row. `updated_at` is an ISO8601 string.
    pub fn new(
        drawer_id: impl Into<String>,
        rating: f64,
        contests: i64,
        updated_at: impl Into<String>,
    ) -> Self {
        Self {
            drawer_id: drawer_id.into(),
            rating,
            contests,
            updated_at: updated_at.into(),
        }
    }
}

/// The rating ledger's schema, declared through PersistenceKit's schema
/// ladder so the DDL is identical on every backend. The name, version and
/// column set match the declaration the 1.8 → 1.9 estate-format capsule
/// applies on populated estates, so `migrate` is a no-op once either has
/// created the table.
pub fn recall_ratings_schema() -> SchemaDeclaration {
    SchemaDeclaration::new(
        "GLKRecallRatings",
        1,
        vec![TableDeclaration::new(
            RECALL_RATINGS_TABLE,
            vec![
                ColumnDeclaration::text("drawer_id"),
                ColumnDeclaration::float("rating"),
                ColumnDeclaration::int("contests"),
                ColumnDeclaration::timestamp("updated_at"),
            ],
            vec!["drawer_id".to_string()],
        )],
    )
}
